### ISSUE TITLE
Memory leaks in beaver

### ISSUE THREAD

**Issue by skotchio** [python-beaver/python-beaver#186]
I'm just started using of beaver on my production servers. I use upstart (https://github.com/josegonzalez/beaver/blob/master/contrib/beaver-upstart.conf) script for running  beaver in background mode. The first time I just started beaver it took around 2% of RAM memory but after 3-5 days it takes 5% - 10% - 19%. It seems there is memory leaks. How can I detect and  fix them?

**Comment by soarpenguin** [comment]
@skotchio have solved the problem yet? 

+++++++++++++++++++++++++++++++++++++
Folks, today I'm faced with the following problem. I've noticed that my logs don't ships to my redis broker and I saw beaver logs and noticed the following records:

Worker has exceeded refresh limit. Terminating
[2013-10-25 23:02:17,168] INFO    Worker process cleanup in progress...
[2013-10-25 23:02:17,172] INFO    Starting worker...
[2013-10-25 23:02:17,173] INFO    Working...
[2013-10-25 23:02:17,176] INFO    Starting queue consumer
[2013-10-25 23:02:17,173] INFO    Working...
[2013-10-25 23:02:17,176] INFO    Starting queue consumer
[2013-10-25 23:02:17,185] INFO    [/var/log/auth.log] - watching logfile
[2013-10-25 23:02:17,190] INFO    [/var/log/nginx/access.log] - watching logfile

[2013-10-25 23:03:22,205] INFO    Queue timeout of "60" seconds exceeded, stopping queue
[2013-10-25 23:03:27,188] INFO    Starting queue consumer 
[2013-10-25 23:04:32,205] INFO    Queue timeout of "60" seconds exceeded, stopping queue
[2013-10-25 23:04:37,196] INFO    Starting queue consumer
[2013-10-25 23:05:42,214] INFO    Queue timeout of "60" seconds exceeded, stopping queue
 . . . . .. 

The previous records repeated many many times. I had to restart beaver process to get it work.
What's problem?
+++++++++++++++++++++++++

I'm faced with above problem yet. thank you!  
$ /opt/python2.7/python beaver.py -v
Beaver 33.0.0

**Comment by skotchio** [comment]
@soarpenguin yes unfortunately I have the same problem yet. I had to write cron task to restart beaver every night and it's seems work,

**Comment by soarpenguin** [comment]
@skotchio  your config have set following option?
   format: rawjson

I have two other host not set above option is work fine.

**Comment by skotchio** [comment]
@soarpenguin yes I have this one. Some later I'll try without it

**Comment by soarpenguin** [comment]
@josegonzalez   

 # time in seconds from last command sent before a queue kills itself   
'queue_timeout': '60',  

what's the mean of the option of queue_timeout?  
and update the "last_update_time" only command is callback?   

 "queue kills itself": you mean have to reconstuct queue?

**Comment by josegonzalez** [comment]
It means the queue hasn't received a message in N seconds. The `last_update_time` *should* be updated every time you send a metric. Honestly, I don't know how to replicate this, and if someone wants to give me access to a system (virtualmachine or otherwise) that I can reliably replicate the issue on, that would be *wonderful*.

**Comment by soarpenguin** [comment]
@josegonzalez  sorry for i can't give you access to a system.

**Comment by soarpenguin** [comment]
@josegonzalez  

if it's possible for queue(size 100) full of addglob command, so beaver can't input log data to queue, can't send data to redis(can't update last_update_time)?

so log of beaver are all of the 
      Queue timeout of "60" seconds exceeded, stopping queue   
and
      Starting queue consumer
can't send log data to redis.

thank you for you reply.

**Comment by soarpenguin** [comment]
@skotchio  what about you log count per hour when beaver collected?

@josegonzalez  i found a phenomenon, when the beaver watched log line count per hour greater a threadhold, it's work fine.    
if the log line count per hour is small, it will  faceed with problem:   logs don't ships to my redis broker and I saw beaver logs and noticed the following records:

[2013-10-25 23:03:22,205] INFO Queue timeout of "60" seconds exceeded, stopping queue
[2013-10-25 23:03:27,188] INFO Starting queue consumer
[2013-10-25 23:04:32,205] INFO Queue timeout of "60" seconds exceeded, stopping queue
[2013-10-25 23:04:37,196] INFO Starting queue consumer
[2013-10-25 23:05:42,214] INFO Queue timeout of "60" seconds exceeded, stopping queue

thank you.

**Comment by soarpenguin** [comment]
@josegonzalez  the log display below issue is right.
if it's possible for queue(size 100) full of addglob command, so beaver can't input log data to queue, can't send data to redis(can't update last_update_time)?

log just like that:   before the time of "2015-03-20 10:13:55" beaver works fine,
                            after "2015-03-20 10:13:55" the queue size is plus, until the size is full.

      some problem let the consumer is not work.

[2015-03-20 10:13:13,020] INFO    Queue timeout of "60 1426817532" seconds exceeded, stopping queue
[2015-03-20 10:13:17,398] INFO    Starting queue consumer
[2015-03-20 10:13:17,505] INFO    Queue size of "0"
[2015-03-20 10:13:19,020] INFO    Queue size of "1"
[2015-03-20 10:13:19,020] INFO    Queue size of "0"
[2015-03-20 10:13:24,020] INFO    Queue size of "0"
[2015-03-20 10:13:29,020] INFO    Queue size of "0"
[2015-03-20 10:13:34,020] INFO    Queue size of "0"
[2015-03-20 10:13:35,020] INFO    Queue size of "1"
[2015-03-20 10:13:35,020] INFO    Queue size of "0"
[2015-03-20 10:13:40,019] INFO    Queue size of "0"
[2015-03-20 10:13:45,019] INFO    Queue size of "0"
[2015-03-20 10:13:50,019] INFO    Queue size of "0"
[2015-03-20 10:13:51,020] INFO    Queue size of "1"
[2015-03-20 10:13:51,020] INFO    Queue size of "0"
[2015-03-20 10:13:55,824] INFO    Worker has exceeded refresh limit. Terminating process...
[2015-03-20 10:13:55,825] INFO    Worker process cleanup in progress...
[2015-03-20 10:13:55,829] INFO    Starting worker...
[2015-03-20 10:13:55,832] INFO    Working...
[2015-03-20 10:13:55,836] INFO    [6802g10288] - watching logfile /var/log/messages
[2015-03-20 10:13:55,837] INFO    [6802g102b5] - watching logfile /var/log/secure
[2015-03-20 10:13:55,839] INFO    Starting queue consumer
[2015-03-20 10:13:55,945] INFO    Queue size of "3"
[2015-03-20 10:14:00,947] INFO    Queue size of "3"
[2015-03-20 10:14:05,949] INFO    Queue size of "3"
[2015-03-20 10:14:10,951] INFO    Queue size of "3"
[2015-03-20 10:14:15,953] INFO    Queue size of "5"
[2015-03-20 10:14:20,955] INFO    Queue size of "5"
[2015-03-20 10:14:25,958] INFO    Queue size of "5"
[2015-03-20 10:14:30,959] INFO    Queue size of "7"
[2015-03-20 10:14:35,961] INFO    Queue size of "7"
[2015-03-20 10:14:40,963] INFO    Queue size of "7"
[2015-03-20 10:14:45,965] INFO    Queue size of "9"
[2015-03-20 10:14:50,967] INFO    Queue size of "9"
[2015-03-20 10:14:55,969] INFO    Queue size of "9"
[2015-03-20 10:15:00,971] INFO    Queue timeout of "60 1426817635" seconds exceeded, stopping queue
[2015-03-20 10:15:05,857] INFO    Starting queue consumer
[2015-03-20 10:15:05,967] INFO    Queue size of "11"
[2015-03-20 10:15:10,969] INFO    Queue size of "11"

................

[2015-03-20 10:26:01,165] INFO    Queue timeout of "60 1426818296" seconds exceeded, stopping queue
[2015-03-20 10:26:06,029] INFO    Starting queue consumer
[2015-03-20 10:26:06,135] INFO    Queue size of "94"
[2015-03-20 10:26:11,138] INFO    Queue size of "96"
[2015-03-20 10:26:16,140] INFO    Queue size of "96"
[2015-03-20 10:26:21,142] INFO    Queue size of "96"
[2015-03-20 10:26:26,144] INFO    Queue size of "96"
[2015-03-20 10:26:31,146] INFO    Queue size of "98"
[2015-03-20 10:26:36,148] INFO    Queue size of "98"
[2015-03-20 10:26:41,150] INFO    Queue size of "98"
[2015-03-20 10:26:46,152] INFO    Queue size of "100"
[2015-03-20 10:26:51,154] INFO    Queue size of "100"
[2015-03-20 10:26:56,156] INFO    Queue size of "100"
[2015-03-20 10:27:01,158] INFO    Queue size of "100"
[2015-03-20 10:27:06,160] INFO    Queue size of "100"
[2015-03-20 10:27:11,162] INFO    Queue timeout of "60 1426818366" seconds exceeded, stopping queue
[2015-03-20 10:27:16,045] INFO    Starting queue consumer
[2015-03-20 10:27:16,152] INFO    Queue size of "100"
[2015-03-20 10:27:21,155] INFO    Queue size of "100"
[2015-03-20 10:27:26,157] INFO    Queue size of "100"
[2015-03-20 10:27:31,159] INFO    Queue size of "100"
[2015-03-20 10:27:36,161] INFO    Queue size of "100"
[2015-03-20 10:27:41,165] INFO    Queue size of "100"
[2015-03-20 10:27:46,167] INFO    Queue size of "100"

**Comment by soarpenguin** [comment]
@josegonzalez    i  add code in beaver/run_queue.py like this:

33             if int(time.time()) - last_update_time > queue_timeout:
 34                 logger.info('Queue timeout of "{0} {1}" seconds exceeded, stopping queue'.format(queue_timeout, last_update_time))
 35 #                pprint(vars(queue))
 36                 break
 37             else:
 38                 logger.info('Queue size of "{0}"'.format(queue.qsize()))

**Comment by soarpenguin** [comment]
@josegonzalez
 
got the problem:
please look at the url: http://bugs.python.org/issue20527
when use the  multiprocessing.Queue,  may deadlocks after “consumer” process death.

**Comment by josegonzalez** [comment]
:(

Thank you for looking into it.

**Comment by josegonzalez** [comment]
Odd, I don't think we ever terminate the queue consumer, so that shouldn't be the issue? Can you check the codebase to see if we manually terminate the consumer?

I wonder if we can just reset the `_rlock` to manually hack around the issue?

**Comment by soarpenguin** [comment]
@josegonzalez
i found the same log like:  http://bugs.python.org/issue20527

[2015-03-19 16:03:40,533] INFO    Queue timeout of "60:1426752155" seconds exceeded, stopping queue
{'_buffer': deque([]),
 '_close': None,
 '_closed': False,
 '_joincancelled': False,
 '_jointhread': None,
 '_maxsize': 20,
 '_notempty': <Condition(<thread.lock object at 0x2b0bea63b610>, 0)>,
 '_opid': 15593,
 '_poll': <built-in method poll of _multiprocessing.Connection object at 0x7684be0>,
 '_reader': <read-only Connection, handle 7>,
 '_recv': <built-in method recv of _multiprocessing.Connection object at 0x7684be0>,

 '_rlock': <Lock(owner=SomeOtherProcess)>,        # this line

 '_sem': <BoundedSemaphore(value=0, maxvalue=20)>,
 '_send': <built-in method send of _multiprocessing.Connection object at 0x7685010>,
 '_thread': None,
 '_wlock': <Lock(owner=None)>,
 '_writer': <write-only Connection, handle 8>}

**Comment by soarpenguin** [comment]
have a solve solution:
in file beaver/run_queue.py:

    def term(signum, frame):
        logger.info("consumer terminating xxxxxxxxxx")
        logger.info('the queue rlock is: {0}'.format(queue._rlock))

\#    signal.signal(signal.SIGTERM, signal.SIG_DFL)
    signal.signal(signal.SIGTERM, term)

catch the signal.SIGTERM, do nothing is OK.

--
##### it's not always have problem in process stopped in signal SIGTERM;
##### if the process catch signal.SIGTERM and owner the _rlock, it must be a problem.

[2015-03-24 08:33:57,134] INFO    Worker has exceeded refresh limit. Terminating process...
[2015-03-24 08:33:57,134] INFO    Worker process cleanup in progress...
[2015-03-24 08:33:57,135] INFO    consumer terminating xxxxxxxxxx
##### [2015-03-24 08:33:57,135] INFO    the queue rlock is: \<Lock(owner=None)\>
[2015-03-24 08:33:57,138] INFO    Starting worker...
[2015-03-24 08:33:57,141] INFO    Working...
[2015-03-24 08:33:57,147] INFO    [802g1bcf0b] - watching logfile /var/log/nginx/access.log
[2015-03-24 08:33:57,148] INFO    [802g18d7c2] - watching logfile /var/log/messages
[2015-03-24 08:33:57,149] INFO    [802g18d7f3] - watching logfile /var/log/secure
\--
[2015-03-24 08:38:57,146] INFO    Worker has exceeded refresh limit. Terminating process...
[2015-03-24 08:38:57,147] INFO    Worker process cleanup in progress...
[2015-03-24 08:38:57,147] INFO    consumer terminating xxxxxxxxxx
##### [2015-03-24 08:38:57,148] INFO    the queue rlock is: \<Lock(owner=Process-129:1)\>
[2015-03-24 08:38:57,151] INFO    Starting worker...
[2015-03-24 08:38:57,153] INFO    Working...
[2015-03-24 08:38:57,160] INFO    [802g1bcf0b] - watching logfile /var/log/nginx/access.log
[2015-03-24 08:38:57,161] INFO    [802g18d7c2] - watching logfile /var/log/messages
[2015-03-24 08:38:57,162] INFO    [802g18d7f3] - watching logfile /var/log/secure

**Comment by josegonzalez** [comment]
Can you make a PR for that?

**Comment by soarpenguin** [comment]
@josegonzalez  sorry, what's the mean of PR?

**Comment by josegonzalez** [comment]
Pull request

**Comment by soarpenguin** [comment]
@josegonzalez  #303

**Comment by JamieCressey** [comment]
We ship a sizeable amount of messages through Beaver per day and havent come across this, can anyone still replicate this with the latest release?

**Comment by geekpete** [comment]
Puka is a nice alternative library to Pika and the RabbitMQ devs recommend it:
https://www.rabbitmq.com/blog/2011/07/08/puka-rethinking-amqp-clients/

**Comment by josegonzalez** [comment]
I don't think that is related to this issue, and the linked article is pretty old. All current rabbitmq documentation uses pika afaik.

**Comment by geekpete** [comment]
Yeah Puka seems to be virtually dead now, few hundred downloads a day vs 20k/day on PyPi for Pika.

Well my experience is that beaver (latest versions) works for a day or two then just stops sending logs. I'm doing about 100meg/day and am only ingesting a single log file per server.
